File: packages/julia/src/Seguranca/Ataques/AtaqueConfusaoCampos.jl
"""
O token é integridade[32] . ciphertext[n] . iv[16]. Se o parser for
ambíguo, um atacante pode reordenar/deslocar os campos e construir um
token que sistemas diferentes interpretam de formas diferentes. Todos os
rearranjos devem ser rejeitados pela integridade.
"""
struct AtaqueConfusaoCampos <: Ataque end

nome(::AtaqueConfusaoCampos) = "Confusão de campos do token (reordenação/deslocamento)"

function executar(a::AtaqueConfusaoCampos, alvo::AlvoCriptografico)
    if !(alvo isa CriptografiaAlvo)
        throw(SkipAtaqueException("Precisa de base64url_encode/decode do alvo."))
    end

    texto = "MENSAGEM_PARA_TESTE_DE_CAMPOS"
    token = encrypt(alvo, texto)
    pref = prefixo(alvo)
    campos = decompor(alvo, base64url_decode(alvo, token[length(pref)+1:end]))

    integridade = campos.integridade
    ciphertext = campos.ciphertext
    iv = campos.iv

    x = UInt8[0x58] # 'X'

    variantes = Pair{String,Vector{UInt8}}[
        "iv no início" => vcat(iv, integridade, ciphertext),
        "ciphertext antes da integridade" => vcat(ciphertext, integridade, iv),
        "iv duplicado no fim" => vcat(integridade, ciphertext, iv, iv),
        "integridade encurtada" => vcat(integridade[2:end], ciphertext, iv),
        "byte extra no início" => vcat(x, integridade, ciphertext, iv),
        "byte extra entre integridade e ciphertext" =>
            vcat(integridade, x, ciphertext, iv),
        "byte extra antes do iv" => vcat(integridade, ciphertext, x, iv),
        "iv rotacionado" => vcat(integridade, ciphertext, reverse(iv)),
        "iv e último byte do ciphertext trocados" =>
            vcat(integridade, ciphertext[1:end-1], iv, ciphertext[end:end]),
    ]

    aceitas = String[]
    for (nome_variante, bruto) in variantes
        token_variante = pref * base64url_encode(alvo, bruto)
        try
            r = decrypt(alvo, token_variante)
            push!(
                aceitas,
                "$nome_variante -> aceito (retornou $(length(codeunits(r))) bytes)",
            )
        catch
            # esperado
        end
    end

    if !isempty(aceitas)
        return ResultadoAtaque(
            nome(a),
            true,
            "critica",
            "$(length(aceitas)) rearranjo(s) de campo foram aceitos",
            Dict{String,Any}("exemplos" => aceitas),
        )
    end

    return ResultadoAtaque(
        nome(a),
        false,
        "info",
        "Todos os $(length(variantes)) rearranjos de campo foram rejeitados",
    )
end

File: packages/typescript/src/Seguranca/Ataques/AtaqueConfusaoCampos.ts
import type { AlvoCriptografico } from '../AlvoCriptografico.ts';
import type { AtaqueInterface } from '../AtaqueInterface.ts';
import { CriptografiaAlvo } from '../CriptografiaAlvo.ts';
import { ResultadoAtaque } from '../ResultadoAtaque.ts';
import { SkipAtaqueException } from '../SkipAtaqueException.ts';

/**
 * O token é integridade[32] . ciphertext[n] . iv[16]. Se o parser for
 * ambíguo, um atacante pode reordenar/deslocar os campos e construir um
 * token que sistemas diferentes interpretam de formas diferentes. Todos os
 * rearranjos devem ser rejeitados pela integridade.
 */
export class AtaqueConfusaoCampos implements AtaqueInterface {
  nome(): string {
    return 'Confusão de campos do token (reordenação/deslocamento)';
  }

  executar(alvo: AlvoCriptografico): ResultadoAtaque {
    if (!(alvo instanceof CriptografiaAlvo)) {
      throw new SkipAtaqueException('Precisa de base64url_encode/decode do alvo.');
    }

    const texto = 'MENSAGEM_PARA_TESTE_DE_CAMPOS';
    const token = alvo.encrypt(texto);
    const prefixo = alvo.prefixo();
    const campos = alvo.decompor(alvo.base64urlDecode(token.slice(prefixo.length)));

    const integridade = campos.integridade;
    const ciphertext = campos.ciphertext;
    const iv = campos.iv;

    const variantes: Record<string, Buffer> = {
      'iv no início': Buffer.concat([iv, integridade, ciphertext]),
      'ciphertext antes da integridade': Buffer.concat([ciphertext, integridade, iv]),
      'iv duplicado no fim': Buffer.concat([integridade, ciphertext, iv, iv]),
      'integridade encurtada': Buffer.concat([integridade.subarray(1), ciphertext, iv]),
      'byte extra no início': Buffer.concat([Buffer.from('X'), integridade, ciphertext, iv]),
      'byte extra entre integridade e ciphertext': Buffer.concat([
        integridade,
        Buffer.from('X'),
        ciphertext,
        iv,
      ]),
      'byte extra antes do iv': Buffer.concat([integridade, ciphertext, Buffer.from('X'), iv]),
      'iv rotacionado': Buffer.concat([integridade, ciphertext, Buffer.from(iv).reverse()]),
      'iv e último byte do ciphertext trocados': Buffer.concat([
        integridade,
        ciphertext.subarray(0, ciphertext.length - 1),
        iv,
        ciphertext.subarray(ciphertext.length - 1),
      ]),
    };

    const aceitas: string[] = [];
    for (const nome of Object.keys(variantes)) {
      const bruto = variantes[nome];
      const tokenVariante = prefixo + alvo.base64urlEncode(bruto);
      try {
        const r = alvo.decrypt(tokenVariante);
        aceitas.push(`${nome} -> aceito (retornou ${Buffer.byteLength(r, 'utf8')} bytes)`);
      } catch (e) {
        // esperado
      }
    }

    if (aceitas.length > 0) {
      return new ResultadoAtaque(
        this.nome(),
        true,
        'critica',
        `${aceitas.length} rearranjo(s) de campo foram aceitos`,
        { exemplos: aceitas },
      );
    }

    return new ResultadoAtaque(
      this.nome(),
      false,
      'info',
      `Todos os ${Object.keys(variantes).length} rearranjos de campo foram rejeitados`,
    );
  }
}
